File: framec_tests/csharp/Match/MatchTest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using NUnit.Framework;

namespace Match
{
    class MatchController : Match
    {
        public MatchController() : base()
        {
        }
        public void log_do(string msg)
        {
            this.tape.Add(msg);
        }
    }
    [TestFixture]
    public class MatchTest
    {
        [Test]
        public void TestEmptyString()
        {
            MatchController sm = new MatchController();
            sm.Empty();
            sm.OnString("");
            Assert.That(sm.tape, Is.EqualTo(new List<string> { "empty" }));

            sm.tape = new List<string>();
            sm.OnString("hi");
            Assert.That(sm.tape, Is.EqualTo(new List<string> { "?" }));
        }

        [Test]
        public void TestIntegerMatch()
        {
            MatchController sm = new MatchController();
            sm.Simple();
            sm.OnInt(0);
            Assert.That(sm.tape, Is.EqualTo(new List<string> { "0" }));

            sm.tape = new List<string>();
            sm.OnInt(42);
            Assert.That(sm.tape, Is.EqualTo(new List<string> { "42" }));

            sm.tape = new List<string>();
            sm.OnInt(-200);
            Assert.That(sm.tape, Is.EqualTo(new List<string> { "-200" }));
        }

        [Test]
        public void TestStringMatch()
        {
            MatchController sm = new MatchController();
            sm.Simple();
            sm.OnString("hello");
            Assert.That(sm.tape, Is.EqualTo(new List<string> { "hello" }));

            sm.tape = new List<string>();
            sm.OnString("goodbye");
            Assert.That(sm.tape, Is.EqualTo(new List<string> { "goodbye" }));

            sm.tape = new List<string>();
            sm.OnString("Testing 1, 2, 3...");
            Assert.That(sm.tape, Is.EqualTo(new List<string> { "testing" }));

            sm.tape = new List<string>();
            sm.OnString("$10!");
            Assert.That(sm.tape, Is.EqualTo(new List<string> { "money" }));

            sm.tape = new List<string>();
            sm.OnString("missing");
            Assert.That(sm.tape, Is.EqualTo(new List<string> { "?" }));

            sm.tape = new List<string>();
            sm.OnString("Testing");
            Assert.That(sm.tape, Is.EqualTo(new List<string> { "?" }));

            sm.tape = new List<string>();
            sm.OnString("");
            Assert.That(sm.tape, Is.EqualTo(new List<string> { "?" }));
        }

        [Test]
        public void TestIntegerMultiMatch()
        {
            MatchController sm = new MatchController();
            sm.Multi();
            sm.OnInt(3);
            Assert.That(sm.tape, Is.EqualTo(new List<string> { "3|-7" }));

            sm.tape = new List<string>();
            sm.OnInt(-7);
            Assert.That(sm.tape, Is.EqualTo(new List<string> { "3|-7" }));

            sm.tape = new List<string>();
            sm.OnInt(-4);
            Assert.That(sm.tape, Is.EqualTo(new List<string> { "-4|5|6" }));

            sm.tape = new List<string>();
            sm.OnInt(5);
            Assert.That(sm.tape, Is.EqualTo(new List<string> { "-4|5|6" }));

            sm.tape = new List<string>();
            sm.OnInt(6);
            Assert.That(sm.tape, Is.EqualTo(new List<string> { "-4|5|6" }));

            sm.tape = new List<string>();
            sm.OnInt(4);
            Assert.That(sm.tape, Is.EqualTo(new List<string> { "?" }));

            sm.tape = new List<string>();
            sm.OnInt(0);
            Assert.That(sm.tape, Is.EqualTo(new List<string> { "?" }));
        }

        [Test]
        public void TestStringMultiMatch()
        {
            var sm = new MatchController();
            sm.Multi();
            sm.OnString("$10");
            Assert.AreEqual(new List<string> { "symbols" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnString("12.5%");
            Assert.AreEqual(new List<string> { "symbols" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnString("@#*!");
            Assert.AreEqual(new List<string> { "symbols" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnString(" ");
            Assert.AreEqual(new List<string> { "whitespace" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnString("  ");
            Assert.AreEqual(new List<string> { "whitespace" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnString("\t");
            Assert.AreEqual(new List<string> { "whitespace" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnString("\n");
            Assert.AreEqual(new List<string> { "whitespace" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnString("10");
            Assert.AreEqual(new List<string> { "?" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnString("#");
            Assert.AreEqual(new List<string> { "?" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnString("   ");
            Assert.AreEqual(new List<string> { "?" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnString("");
            Assert.AreEqual(new List<string> { "?" }, sm.tape);
            sm.tape = new List<string>();
        }

        [Test]
        public void TestIntegerNestedMatch()
        {
            var sm = new MatchController();
            sm.Nested();
            sm.OnInt(1);
            Assert.AreEqual(new List<string> { "1-3", "1" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnInt(2);
            Assert.AreEqual(new List<string> { "1-3", "2" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnInt(3);
            Assert.AreEqual(new List<string> { "1-3", "3" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnInt(4);
            Assert.AreEqual(new List<string> { "4-5", "4" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnInt(5);
            Assert.AreEqual(new List<string> { "4-5", "5" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnInt(10);
            Assert.AreEqual(new List<string> { "too big" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnInt(0);
            Assert.AreEqual(new List<string> { "too small" }, sm.tape);
        }

        [Test]
        public void TestStringNestedMatch()
        {
            var sm = new MatchController();
            sm.Nested();
            sm.OnString("hello");
            Assert.AreEqual(new List<string> { "greeting", "English" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnString("hola");
            Assert.AreEqual(new List<string> { "greeting", "Spanish" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnString("bonjour");
            Assert.AreEqual(new List<string> { "greeting", "French" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnString("goodbye");
            Assert.AreEqual(new List<string> { "farewell", "English" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnString("adios");
            Assert.AreEqual(new List<string> { "farewell", "Spanish" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnString("au revoir");
            Assert.AreEqual(new List<string> { "farewell", "French" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnString("hallo");
            Assert.AreEqual(new List<string> { "?" }, sm.tape);
            sm.tape = new List<string>();
            sm.OnString("ciao");
            Assert.AreEqual(new List<string> { "?" }, sm.tape);
        }

        [Test]
        public void TestIntegerHierarchicalMatch()
        {
            var sm = new MatchController();
            sm.Child();
            sm.OnInt(0);
            Assert.AreEqual("6", sm.state_info());
            Assert.AreEqual(0, sm.tape.Count);

            sm = new MatchController();
            sm.Child();
            sm.OnInt(4);
            Assert.AreEqual("5", sm.state_info());
            Assert.AreEqual(new List<string> { "4" }, sm.tape);

            sm.tape = new List<string>();
            sm.OnInt(5);
            Assert.AreEqual("6", sm.state_info());
            Assert.AreEqual(new List<string> { "5" }, sm.tape);

            sm = new MatchController();
            sm.Child();
            sm.OnInt(5);
            Assert.AreEqual("6", sm.state_info());
            Assert.AreEqual(new List<string> { "5" }, sm.tape);

            sm = new MatchController();
            sm.Child();
            sm.OnInt(3);
            Assert.AreEqual("5", sm.state_info());
            Assert.AreEqual(new List<string> { "3", "?" }, sm.tape);

            sm.tape = new List<string>();
            sm.OnInt(42);
            Assert.AreEqual("5", sm.state_info());
            Assert.AreEqual(new List<string> { "42 in child", "42" }, sm.tape);

            sm.tape = new List<string>();
            sm.OnInt(-200);
            Assert.AreEqual("5", sm.state_info());
            Assert.AreEqual(new List<string> { "no match in child", "-200" }, sm.tape);

            sm.tape = new List<string>();
            sm.OnInt(100);
            Assert.AreEqual("5", sm.state_info());
            Assert.AreEqual(new List<string> { "no match in child", "?" }, sm.tape);
        }

        [Test]
        public void TestStringHierarchicalMatch()
        {
            MatchController sm = new MatchController();
            sm.Child();
            sm.OnString("goodbye");
            Assert.AreEqual("6", sm.state_info());
            Assert.AreEqual(0, sm.tape.Count);

            sm = new MatchController();
            sm.Child();
            sm.OnString("hello");
            Assert.AreEqual("5", sm.state_info());
            Assert.AreEqual(new List<string> { "hello in child", "hello" }, sm.tape);

            sm.tape = new List<string>();
            sm.OnString("Testing 1, 2, 3...");
            Assert.AreEqual("5", sm.state_info());
            Assert.AreEqual(new List<string> { "testing in child" }, sm.tape);

            sm.tape = new List<string>();
            sm.OnString("$10!");
            Assert.AreEqual("5", sm.state_info());
            Assert.AreEqual(new List<string> { "no match in child", "money" }, sm.tape);

            sm.tape = new List<string>();
            sm.OnString("testing 1, 2, 3...");
            Assert.AreEqual("5", sm.state_info());
            Assert.AreEqual(new List<string> { "no match in child", "?" }, sm.tape);
        }



    }
}

File: framec_tests/cpp/match/match_test.cpp
#include <gtest/gtest.h>
#include <vector>
#include <string>
#include "match.cpp"

class MatchController : public Match {
public:
    MatchController() : Match() {}
};

class MatchTest : public ::testing::Test {
protected:
    void SetUp() override {
        sm = new MatchController();
    }

    void TearDown() override {
        delete sm;
    }

    MatchController* sm;
};

TEST_F(MatchTest, TestEmptyString) {
    sm->Empty();
    sm->OnString("");
    std::vector<std::string> expected;
    expected.push_back("empty");
    ASSERT_EQ(sm->tape, expected);
    
    sm->tape.clear();
    sm->OnString("hi");
    std::vector<std::string> expected1;
    expected1.push_back("?");
    ASSERT_EQ(sm->tape, expected1);
}

TEST_F(MatchTest, TestIntegerMatch) {
    sm->Simple();
    sm->OnInt(0);
    std::vector<std::string> expected;
    expected.push_back("0");
    ASSERT_EQ(sm->tape, expected);
    
    sm->tape.clear();
    sm->OnInt(42);
    std::vector<std::string> expected1;
    expected1.push_back("42");
    ASSERT_EQ(sm->tape, expected1);
    
    sm->tape.clear();
    sm->OnInt(-200);
    std::vector<std::string> expected2;
    expected2.push_back("-200");
    ASSERT_EQ(sm->tape, expected2);
}

TEST_F(MatchTest, TestStringMatch) {
    sm->Simple();
    sm->OnString("hello");
    std::vector<std::string> expected;
    expected.push_back("hello");
    ASSERT_EQ(sm->tape, expected);

    sm->tape.clear();
    sm->OnString("goodbye");
    std::vector<std::string> expected1;
    expected1.push_back("goodbye");
    ASSERT_EQ(sm->tape, expected1);

    sm->tape.clear();
    sm->OnString("Testing 1, 2, 3...");
    std::vector<std::string> expected2;
    expected2.push_back("testing");
    ASSERT_EQ(sm->tape, expected2);
    std::vector<std::string> expected3;
    expected3.push_back("testing");
    ASSERT_EQ(sm->tape, expected3);

    sm->tape.clear();
    sm->OnString("$10!");
    std::vector<std::string> expected4;
    expected4.push_back("money");
    ASSERT_EQ(sm->tape, expected4);
    std::vector<std::string> expected5;
    expected5.push_back("money");
    ASSERT_EQ(sm->tape, expected5);

    sm->tape.clear();
    sm->OnString("missing");
    std::vector<std::string> expected6;
    expected6.push_back("?");
    ASSERT_EQ(sm->tape, expected6);
    std::vector<std::string> expected7;
    expected7.push_back("?");
    ASSERT_EQ(sm->tape, expected7);

    sm->tape.clear();
    sm->OnString("Testing");
    std::vector<std::string> expected8;
    expected8.push_back("?");
    ASSERT_EQ(sm->tape, expected8);
    std::vector<std::string> expected9;
    expected9.push_back("?");
    ASSERT_EQ(sm->tape, expected9);

    sm->tape.clear();
    sm->OnString("");
    std::vector<std::string> expected10;
    expected10.push_back("?");
    ASSERT_EQ(sm->tape, expected10);
    std::vector<std::string> expected11;
    expected11.push_back("?");
    ASSERT_EQ(sm->tape, expected11);
}

TEST_F(MatchTest, TestIntegerMultiMatch) {
    sm->Multi();
    sm->OnInt(3);
    std::vector<std::string> expected;
    expected.push_back("3|-7");
    ASSERT_EQ(sm->tape, expected);

    sm->tape.clear();
    sm->OnInt(-7);
    std::vector<std::string> expected1;
    expected1.push_back("3|-7");
    ASSERT_EQ(sm->tape, expected1);

    sm->tape.clear();
    sm->OnInt(-4);
    std::vector<std::string> expected2;
    expected2.push_back("-4|5|6");
    ASSERT_EQ(sm->tape, expected2);

    sm->tape.clear();
    sm->OnInt(5);
    std::vector<std::string> expected3;
    expected3.push_back("-4|5|6");
    ASSERT_EQ(sm->tape, expected3);

    sm->tape.clear();
    sm->OnInt(6);
    std::vector<std::string> expected4;
    expected4.push_back("-4|5|6");
    ASSERT_EQ(sm->tape, expected4);

    sm->tape.clear();
    sm->OnInt(4);
    std::vector<std::string> expected5;
    expected5.push_back("?");
    ASSERT_EQ(sm->tape, expected5);

    sm->tape.clear();
    sm->OnInt(0);
    std::vector<std::string> expected6;
    expected6.push_back("?");
    ASSERT_EQ(sm->tape, expected6);
}

TEST_F(MatchTest, TestStringMultiMatch) {
    sm->Multi();
    sm->OnString("$10");
    std::vector<std::string> expected;
    expected.push_back("symbols");
    ASSERT_EQ(sm->tape, expected);

    sm->tape.clear();
    sm->OnString("12.5%");
    std::vector<std::string> expected1;
    expected1.push_back("symbols");
    ASSERT_EQ(sm->tape, expected1);

    sm->tape.clear();
    sm->OnString("@#*!");
    std::vector<std::string> expected2;
    expected2.push_back("symbols");
    ASSERT_EQ(sm->tape, expected2);

    sm->tape.clear();
    sm->OnString(" ");
    std::vector<std::string> expected3;
    expected3.push_back("whitespace");
    ASSERT_EQ(sm->tape, expected3);

    sm->tape.clear();
    sm->OnString(" ");
    std::vector<std::string> expected4;
    expected4.push_back("whitespace");
    ASSERT_EQ(sm->tape, expected4);

    sm->tape.clear();
    sm->OnString("\t");
    std::vector<std::string> expected5;
    expected5.push_back("whitespace");
    ASSERT_EQ(sm->tape, expected5);

    sm->tape.clear();
    sm->OnString("\n");
    std::vector<std::string> expected6;
    expected6.push_back("whitespace");
    ASSERT_EQ(sm->tape, expected6);

    sm->tape.clear();
    sm->OnString("10");
    std::vector<std::string> expected7;
    expected7.push_back("?");
    ASSERT_EQ(sm->tape, expected7);

    sm->tape.clear();
    sm->OnString("#");
    std::vector<std::string> expected8;
    expected8.push_back("?");
    ASSERT_EQ(sm->tape, expected8);

    sm->tape.clear();
    sm->OnString("   ");
    std::vector<std::string> expected9;
    expected9.push_back("?");
    ASSERT_EQ(sm->tape, expected9);

    sm->tape.clear();
    sm->OnString("");
    std::vector<std::string> expected10;
    expected10.push_back("?");
    ASSERT_EQ(sm->tape, expected10);
}

TEST_F(MatchTest, TestIntegerNestedMatch) {
    sm->Nested();
    sm->OnInt(1);
    std::vector<std::string> expected;
    expected.push_back("1-3");
    expected.push_back("1");
    ASSERT_EQ(sm->tape, expected);

    sm->tape.clear();
    sm->OnInt(2);
    std::vector<std::string> expected1;
    expected1.push_back("1-3");
    expected1.push_back("2");
    ASSERT_EQ(sm->tape, expected1);

    sm->tape.clear();
    sm->OnInt(3);
    std::vector<std::string> expected2;
    expected2.push_back("1-3");
    expected2.push_back("3");
    ASSERT_EQ(sm->tape, expected2);

    sm->tape.clear();
    sm->OnInt(4);
    std::vector<std::string> expected3;
    expected3.push_back("4-5");
    expected3.push_back("4");
    ASSERT_EQ(sm->tape, expected3);

    sm->tape.clear();
    sm->OnInt(5);
    std::vector<std::string> expected4;
    expected4.push_back("4-5");
    expected4.push_back("5");
    ASSERT_EQ(sm->tape, expected4);

    sm->tape.clear();
    sm->OnInt(10);
    std::vector<std::string> expected5;
    expected5.push_back("too big");
    ASSERT_EQ(sm->tape, expected5);

    sm->tape.clear();
    sm->OnInt(0);
    std::vector<std::string> expected6;
    expected6.push_back("too small");
    ASSERT_EQ(sm->tape, expected6);
}

TEST_F(MatchTest, TestStringNestedMatch) {
    sm->Nested();
    sm->OnString("hello");
    std::vector<std::string> expected;
    expected.push_back("greeting");
    expected.push_back("English");
    ASSERT_EQ(sm->tape, expected);

    sm->tape.clear();
    sm->OnString("hola");
    std::vector<std::string> expected1;
    expected1.push_back("greeting");
    expected1.push_back("Spanish");
    ASSERT_EQ(sm->tape, expected1);

    sm->tape.clear();
    sm->OnString("bonjour");
    std::vector<std::string> expected2;
    expected2.push_back("greeting");
    expected2.push_back("French");
    ASSERT_EQ(sm->tape, expected2);

    sm->tape.clear();
    sm->OnString("goodbye");
    std::vector<std::string> expected3;
    expected3.push_back("farewell");
    expected3.push_back("English");
    ASSERT_EQ(sm->tape, expected3);

    sm->tape.clear();
    sm->OnString("adios");
    std::vector<std::string> expected4;
    expected4.push_back("farewell");
    expected4.push_back("Spanish");
    ASSERT_EQ(sm->tape, expected4);

    sm->tape.clear();
    sm->OnString("au revoir");
    std::vector<std::string> expected5;
    expected5.push_back("farewell");
    expected5.push_back("French");
    ASSERT_EQ(sm->tape, expected5);

    sm->tape.clear();
    sm->OnString("hallo");
    std::vector<std::string> expected6;
    expected6.push_back("?");
    ASSERT_EQ(sm->tape, expected6);

    sm->tape.clear();
    sm->OnString("ciao");
    std::vector<std::string> expected7;
    expected7.push_back("?");
    ASSERT_EQ(sm->tape, expected7);
}

TEST_F(MatchTest, TestIntegerHierarchicalMatch) {
    sm->Child();
    sm->OnInt(0);
    ASSERT_EQ(sm->state_info(), "6");
    sm->tape.clear();

    sm = new MatchController();
    sm->Child();
    sm->OnInt(4);
    ASSERT_EQ(sm->state_info(), "5");
    std::vector<std::string> expected;
    expected.push_back("4");
    ASSERT_EQ(sm->tape, expected);

    sm->tape.clear();
    sm->OnInt(5);
    ASSERT_EQ(sm->state_info(), "6");
    std::vector<std::string> expected1;
    expected1.push_back("5");
    ASSERT_EQ(sm->tape, expected1);
    
    sm = new MatchController();
    sm->tape.clear();
    sm->Child();
    sm->OnInt(3);
    ASSERT_EQ(sm->state_info(), "5");
    std::vector<std::string> expected2;
    expected2.push_back("3");
    expected2.push_back("?");
    ASSERT_EQ(sm->tape, expected2);

    sm->tape.clear();
    sm->OnInt(42);
    ASSERT_EQ(sm->state_info(), "5");
    std::vector<std::string> expected3;
    expected3.push_back("42 in child");
    expected3.push_back("42");
    ASSERT_EQ(sm->tape, expected3);

    sm->tape.clear();
    sm->OnInt(-200);
    ASSERT_EQ(sm->state_info(), "5");
    std::vector<std::string> expected4;
    expected4.push_back("no match in child");
    expected4.push_back("-200");
    ASSERT_EQ(sm->tape, expected4);

    sm->tape.clear();
    sm->OnInt(100);
    ASSERT_EQ(sm->state_info(), "5");
    std::vector<std::string> expected5;
    expected5.push_back("no match in child");
    expected5.push_back("?");
    ASSERT_EQ(sm->tape, expected5);
}

TEST_F(MatchTest, TestStringHierarchicalMatch) {
    sm->Child();
    sm->OnString("goodbye");
    ASSERT_EQ(sm->state_info(), "6");
    sm->tape.clear();

    sm = new MatchController();
    sm->Child();
    sm->OnString("hello");
    ASSERT_EQ(sm->state_info(), "5");
    std::vector<std::string> expected;
    expected.push_back("hello in child");
    expected.push_back("hello");
    ASSERT_EQ(sm->tape, expected);

    sm->tape.clear();
    sm->OnString("Testing 1, 2, 3...");
    ASSERT_EQ(sm->state_info(), "5");
    std::vector<std::string> expected1;
    expected1.push_back("testing in child");
    ASSERT_EQ(sm->tape, expected1);

    sm->tape.clear();
    sm->OnString("$10!");
    ASSERT_EQ(sm->state_info(), "5");
    std::vector<std::string> expected2;
    expected2.push_back("no match in child");
    expected2.push_back("money");
    ASSERT_EQ(sm->tape, expected2);

    sm->tape.clear();
    sm->OnString("testing 1, 2, 3...");
    ASSERT_EQ(sm->state_info(), "5");
    std::vector<std::string> expected3;
    expected3.push_back("no match in child");
    expected3.push_back("?");
    ASSERT_EQ(sm->tape, expected3);
}

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
